### ISSUE TITLE
`handle_exit` called with missing `starter` causes a crash

### ISSUE THREAD

**Issue by tomalexhughes** [gleam-lang/otp#67]
Creating an issue for [this problem](https://discord.com/channels/768594524158427167/1265247325073444864/1265247325073444864) discussed in the Discord.

It's possible for the `handle_exit` function in the supervisor module to be called with state that is missing the `starter.exec` function.

This causes the following error due to this [`let assert`](https://github.com/gleam-lang/otp/blob/fba1d59d97ab9a64c674ef8e196b3484d7506ccd/src/gleam/otp/supervisor.gleam#L303)

```gleam
Runtime terminating during boot ({#{function=><<"handle_exit">>,line=>303,message=><<"Assertion pattern match failed">>,module=><<"gleam/otp/supervisor">>,value=>none,gleam_error=>let_assert},[{'gleam@otp@supervisor',handle_exit,2,[{file,"/Users/tom/Developer/gloogle/apps/backend/build/dev/erlang/gleam_otp/_gleam_artefacts/gleam@otp@supervisor.erl"},{line,235}]},{'gleam@otp@actor',loop,1,[{file,"/Users/tom/Developer/gloogle/apps/backend/build/dev/erlang/gleam_otp/_gleam_artefacts/gleam@otp@actor.erl"},{line,150}]}]})
```

Unfortunately I do not have a way to reproduce at the moment and I'm unsure as to the exact details that cause the none value.

**Comment by lpil** [comment]
Thank you.

**Comment by ghivert** [comment]
FYI, I have a regular error on Gloogle due to this error, because I'm exiting normally some children. That triggers server to restart. It's not a big deal at the moment, but it's bothering enough I'd like to fix it. Are you interested in some PR @lpil ? In case you are, do you have some opinions or some hints to get started on the subject?

**Comment by lpil** [comment]
PRs welcome but I've not investigated this.

**Comment by ghivert** [comment]
I think I understood why it's happening: I'm using a supervisor to start fire & forget tasks. In Gloogle, when I want to extract signatures, I'm spawning a process to do it in an async way, but I'm starting it within a supervisor. So when it shutdown, there's no other children to start, because it's supposed to stop. Am I right to say that every supervisor children processes should be thought as long running processes, that restart automatically?

**Comment by lpil** [comment]
> Am I right to say that every supervisor children processes should be thought as long running processes, that restart automatically?

No, it would depend on what supervisor you are using and how it has been configured.
